cpp/src/parquet/column_reader.cc

```diff
@@ -1478,16 +1480,29 @@ class TypedRecordReader : public TypedColumnReaderImpl<DType>,
     // We skipped the levels by incrementing 'levels_position_'. For values
     // we do not have a buffer, so we need to read them and throw them away.
     // First we need to figure out how many present/not-null values there are.
-    std::shared_ptr<::arrow::ResizableBuffer> valid_bits;
-    valid_bits = AllocateBuffer(this->pool_);
-    PARQUET_THROW_NOT_OK(valid_bits->Resize(bit_util::BytesForBits(skipped_records),
-                                            /*shrink_to_fit=*/true));
+    int64_t buffer_size = bit_util::BytesForBits(skipped_records);
+    if (valid_bits_for_skip_ == nullptr) {
+      valid_bits_for_skip_ = AllocateBuffer(this->pool_);
+    }
+    if (buffer_size > valid_bits_for_skip_->size()) {
```

Comment:
Instead of dynamically resizing the buffer to match the number the records, have you tried skipping in chunks?

Reply:
For example, you could skip `kSkipLevelBufferSize * 8` records at a time without ever resizing the intermediate buffer.

Reply:
Good point. Theoretically It should also work. One question is how can the  [skipped_records](https://github.com/apache/arrow/blob/main/cpp/src/parquet/column_reader.cc#L1473)  for `SkipRecordsInBufferNonRepeated` ever exceed [kMinLevelBatchSize](https://github.com/apache/arrow/blob/main/cpp/src/parquet/column_reader.cc#L1369) (i.e., 1024)? If it's always less than 1024, adding a chunk based logic here would be unnecessary (resizing to 0 feels more straightforward and less risky as it follows the old behavior).

Reply:
> One question is how can the [skipped_records](https://github.com/apache/arrow/blob/main/cpp/src/parquet/column_reader.cc#L1473) for `SkipRecordsInBufferNonRepeated` ever exceed [kMinLevelBatchSize](https://github.com/apache/arrow/blob/main/cpp/src/parquet/column_reader.cc#L1369) (i.e., 1024)?

Why not? You can add an assertion to check this with the test suite.

Reply:
Thanks for the review. I tried the tests and was not able to get a case with `skipped_records >=1024`. I may misread the code but it seems the  `skipped_records` [can never exceed ](https://github.com/apache/arrow/blob/main/cpp/src/parquet/column_reader.cc#L1473) the number of buffered  levels. And the number of buffered levels will not exceed   [kMinLevelBatchSize](https://github.com/apache/arrow/blob/main/cpp/src/parquet/column_reader.cc#L1369).

Reply:
Hi @pitrou , could you please take a look at this and let me know your comments? Thanks!

Reply:
Again, I would suggest working in chunks instead of resizing the internal buffer.